### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by PeterPetrik**

### CODE REVIEW

File: src/clientcommands.cpp [hunk 1]
```diff
@@ -106,6 +108,63 @@ void ClientCommands::sync()
...
+  else
+  {
+    suffix = '\n';
+  }
+
+  out << prefix;
+  printVal( isJsonFormat, out, "name", lp.projectName, false );
+  printVal( isJsonFormat, out, "namespace", lp.projectNamespace, false );
+  printVal( isJsonFormat, out, "id", lp.id(), false );
+  printVal( isJsonFormat, out, "localVersion", QString::number( lp.localVersion ), true );
```
Comment (wonder-sk) [inline src/clientcommands.cpp]: Can we do the JSON output properly, through QJsonDocument + other Qt classes? Like this, we would have encoding problems if a value contains double qoutes, we don't support arrays etc.

Not having a single `printVal` function for both JSON and human-oriented output would give us more flexibility how we format output for the user...

File: src/clientmain.cpp [hunk 2]
```diff
@@ -66,18 +73,22 @@ Args parseArgs()
   parser.addOption( passwordOption );
 
   parser.parse( QCoreApplication::arguments() );
-  if ( parser.isSet( helpOption ) )
-    parser.showHelp(); // exits the app
+
+  const QStringList posArgs = parser.positionalArguments();
+  if ( parser.isSet( helpOption ) || posArgs.isEmpty() )
+    parser.showHelp(); // exits the ap
```
Comment (wonder-sk) [inline src/clientmain.cpp]:
```suggestion
    parser.showHelp(); // exits the app
```

File: src/clientcommands.cpp [hunk 3]
```diff
@@ -106,6 +108,63 @@ void ClientCommands::sync()
...
+  out << quote << key << quote << ":"  << space << quote << val << quote
+      ;
+
+  if ( !isLast )
+    out << separator;
+}
+
+void ClientCommands::info( bool isJsonFormat )
+{
+  Q_ASSERT( isAuthorized() );
```
Comment (wonder-sk) [inline src/clientcommands.cpp]: 1. why do we need to check if user is authorized?
2. why assert? better to output a normal error message (probably different syntax whether using JSON or not)

File: src/clientmain.cpp [hunk 4]
```diff
@@ -55,7 +57,12 @@ Args parseArgs()
   parser.setApplicationDescription( "Mergin Command Line Client" );
   QCommandLineOption helpOption = parser.addHelpOption();
   QCommandLineOption versionOption = parser.addVersionOption();
-  parser.addPositionalArgument( "command", "create/remove/download/sync" );
+  QCommandLineOption logOption( "log", "log file (debug output)", "log" );
+  parser.addOption( logOption );
+  QCommandLineOption jsonOption( "json", "output as JSON format (e.g. in info format)", "json" );
+  parser.addOption( jsonOption );
+
+  parser.addPositionalArgument( "command", "create/remove/download/sync/info" );
```
Comment (wonder-sk) [inline src/clientmain.cpp]: Is there any help for these individual commands?

File: src/clientmain.cpp [hunk 5]
```diff
@@ -66,18 +73,22 @@ Args parseArgs()
...
   args.user = parseEnvArg( parser.value( userOption ), "MERGIN_USER" );
   args.pass = parseEnvArg( parser.value( passwordOption ), "MERGIN_PASSWORD" );
+  args.logFile = parser.value( logOption );
+  args.isJsonFormat = parser.isSet( jsonOption );
 
-  const QStringList posArgs = parser.positionalArguments();
-  if ( posArgs.isEmpty() )
-    throw QString( "command is required positional argument" );
+  if ( !args.logFile.isEmpty() )
+    CoreUtils::setLogFilename( args.logFile );
```
Comment (wonder-sk) [inline src/clientmain.cpp]: What happens if the log file is not set?

File: src/clientmain.cpp [hunk 6]
```diff
@@ -66,18 +73,22 @@ Args parseArgs()
...
+    parser.showHelp(); // exits the ap
+
   if ( parser.isSet( versionOption ) )
     parser.showVersion(); // exits the app
 
   args.url = parseEnvArg( parser.value( urlOption ), "MERGIN_URL", false ); // MerginApi has public.cloudmergin.com as default
   args.user = parseEnvArg( parser.value( userOption ), "MERGIN_USER" );
   args.pass = parseEnvArg( parser.value( passwordOption ), "MERGIN_PASSWORD" );
+  args.logFile = parser.value( logOption );
+  args.isJsonFormat = parser.isSet( jsonOption );
```
Comment (wonder-sk) [inline src/clientmain.cpp:116]: This should be valid only for "info" or all commands?

**Comment by PeterPetrik** [comment]
Screen Shot 2021-04-27 at 8 49 34

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**